widgets/secrets/src/UpdateModal.tsx

```diff
@@ -20,33 +23,69 @@ export default function UpdateModal({ open, secret, toolbox, onHide }: UpdateMod
     const { errors, setMessageAsError, clearErrors, setErrors } = useErrors();
     const [canUpdateSecret, enableSecretUpdate, disableSecretUpdate] = useBoolean(true);
     const [secretValue, setSecretValue, clearSecretValue] = useInput('');
+    const [useSecretProvider, setUseSecretProvider] = useState(secret.provider_name !== null);
+    const [secretProvider, setSecretProvider, clearSecretProvider] = useInput('');
+    const [secretProviderPath, setSecretProviderPath, clearSecretProviderPath] = useInput('');
+    const [secretProviders, setSecretProviders] = useState<SecretProvidersWidget.DataItem[]>();
+    const [secretProviderOptions, setSecretProviderOptions] = useState<Record<string, any>>();
 
     useOpenProp(open, () => {
         setLoading();
         enableSecretUpdate();
         clearErrors();
         clearSecretValue();
+        clearSecretProvider();
+        clearSecretProviderPath();
+        fetchSecretProviders();
 
         const actions = new Stage.Common.Secrets.Actions(toolbox.getManager());
         actions
-            .doGet(secret.key)
-            .then(({ is_hidden_value: isHidden, value }) => {
-                clearErrors();
-                setSecretValue(value);
-
-                if (isHidden && isEmpty(value)) {
-                    disableSecretUpdate();
-                } else {
-                    enableSecretUpdate();
+            .doGetAllSecrets()
+            .then(data => {
+                const secretData = data.items.find((item: Secret) => item.key === secret.key);
+                const hasSecretProvider = !isEmpty(secret.provider_name);
+                if (secretData) {
+                    const isHidden = secretData.is_hidden_data;
+                    if (hasSecretProvider) {
+                        setUseSecretProvider(true);
+                        setSecretProvider(secretData.provider_name);
+                        actions.doGet(secretData.key).then(({ provider_options: providerOptions }) => {
+                            setSecretProviderOptions(providerOptions);
+                            setSecretProviderPath(providerOptions!.path);
+                        });
+                    } else {
+                        setUseSecretProvider(false);
+                        actions
+                            .doGet(secretData.key)
+                            .then(({ value }) => {
+                                setSecretValue(value);
+                            })
+                            .catch(err => setErrors({ error: err.message }));
+                    }
+
+                    if (isHidden) {
+                        disableSecretUpdate();
+                    } else {
+                        enableSecretUpdate();
+                    }
                 }
             })
-            .catch(err => setErrors({ secretValue: err.message }))
+            .catch(err => setErrors({ error: err.message }))
```

Comment:
We could use here `setMessageAsError` function as with other `catch` blocks 🎉 

```suggestion
                            .catch(setMessageAsError);
                    }

                    if (isHidden) {
                        disableSecretUpdate();
                    } else {
                        enableSecretUpdate();
                    }
                }
            })
            .catch(setMessageAsError)
```

Reply:
done ✔️  https://github.com/cloudify-cosmo/cloudify-stage/pull/2391/commits/06578b882f48436912415dc8f639ac0f0b18455f